### ISSUE TITLE
Should `\normalcolor` update current color `.`?

### ISSUE THREAD

**Issue by muzimuzhi** [latex3/xcolor#19]
```tex
\documentclass{article}
\usepackage{xcolor}

\begin{document}
\color{red} red
\normalcolor normal
\color{.}   red
\end{document}
```
![image](https://user-images.githubusercontent.com/6376638/181866747-baade9d3-1138-4730-9130-22aa6b95d66b.png)

**Comment by davidcarlisle** [comment]
I'd say yes since it is just `\let\current@color\default@color\set@color`  the intention was to set the current color, basically it's only doing `color` package version though..

**Comment by u-fischer** [comment]
Imho yes too, but the problem is that `\current@color` and `\default@color` contain only only the backend command, they miss the data for the model and the value. To properly setup the `.` color from `\normalcolor` one need additional commands here, and then one would have to check if there are packages which change `\current@color`.

**Comment by davidcarlisle** [comment]
@u-fischer  rather than change `\current@color` I was thinking `\normalcolor` should set `\XC@current@color`  just as if it did `\color{normal-color-name}`

**Comment by u-fischer** [comment]
@davidcarlisle  sure, but to which value? I mean it looks like this:

~~~~
\XC@current@color=macro:
->\xcolor@ {}{1 0 0 rg 1 0 0 RG}{rgb}{1,0,0}.
~~~~

To set this to the default color, it can take `\default@color` to set the stuff in the second brace. But how should it fill the third and fourth brace? One can set it to gray  and 1 from the start, but if someone wants to change the default color e.g. to a red or some spot color it should work too. One need some `\default@color@model` and `default@color@value`. Or a default l3color from which one can extract all this.

**Comment by davidcarlisle** [comment]
@u-fischer  ah I meant something like this (which doesn't work, but should be fixable?)

```
\documentclass{article}
\usepackage{xcolor}
\makeatletter
% current code
% \protected\def\normalcolor{\let\current@color\default@color\set@color}
% \AtBeginDocument{\let\default@color\current@color}
\protected\def\normalcolor{\let\current@color\default@color\set@color
                           \let\XC@current@color\XC@default@color% xc equiv of \se@color here....
                            }
\AtBeginDocument{\let\default@color\current@color\let\XC@default@color\XC@current@color}

\begin{document}


\color{red} red
\normalcolor normal
\color{.}   red
\end{document}
```

**Comment by u-fischer** [comment]
` \let\XC@current@color\XC@default@color\XC@display` should work imho. The question is if there are packages which change `\current@color`.

**Comment by davidcarlisle** [comment]
```
\documentclass{article}
\usepackage{xcolor}
\makeatletter
% current code
% \protected\def\normalcolor{\let\current@color\default@color\set@color}
% \AtBeginDocument{\let\default@color\current@color}
\protected\def\normalcolor{\let\current@color\default@color
                           \let\XC@current@color\XC@default@color\XC@display
                           \set@color}
\AtBeginDocument{\let\default@color\current@color\let\XC@default@color\XC@current@color}
\makeatother

\begin{document}


\color{red} red
\normalcolor normal
\color{.}   red
\end{document}
```

Works,  yes

**Comment by muzimuzhi** [comment]
`\XC@display` calls `\set@color` near end (see line 5888 below), hence the `\set@color` at the end of new `\normalcolor` seems duplicate, which will push current color to the color stack the second time.
https://github.com/latex3/xcolor/blob/849682246582946835d28c8f9b2081ff2c340e09/xcolor.dtx#L5879-L5888

**Comment by u-fischer** [comment]
@muzimuzhi I added code to update the dot color. It is together with more changes in the changekeyhandler branch.  Could you perhaps test it with pgf as you have a much larger testsuite for colors? (If you have tests using colortbl you will perhaps need also the new colortbl which has been sent to ctan yesterday.)

**Comment by muzimuzhi** [comment]
Sorry actually I don't have such a testsuite for colors. The current issue report came from discussions in CTeX-org/forum#263 and the duplicate-color-push is found when I was preparing example for #21.

Currently `pgf-tikz` has only one box-content-based test. I checked that one and three beamer tests in `ctex-kit/ctex` and `beamer` and all the checks passed.

**Comment by u-fischer** [comment]
Fixed in version 3.00. The example now outputs:

![image](https://github.com/latex3/xcolor/assets/4047173/75482b78-825d-4130-876b-84d2aa808508)
